File: DSA/TwoPointers/TwoSum/two_sum.py


# https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/


def two_sum(nums, target):
    # Time: O(n), Space: O(1)
    left = 0
    right = len(nums) - 1

    while left < right:
        curr_sum = nums[left] + nums[right]

        if curr_sum > target:
            right -= 1
        elif curr_sum < target:
            left += 1
        else:
            return [left + 1, right + 1]


if __name__ == '__main__':
    n1 = [2, 7, 11, 15]
    t1 = 9

    n2 = [2, 3, 4]
    t2 = 6

    n3 = [-1, 0]
    t3 = -1

    n4 = [1, 3, 4, 5, 7, 10, 11]
    t4 = 9

    print(two_sum(n1, t1))
    print(two_sum(n2, t2))
    print(two_sum(n3, t3))
    print(two_sum(n4, t4))

File: DSA_Problems/ArrayHashing/TwoSum/TwoSum.java
package DSA_Problems.ArrayHashing.TwoSum;

import java.util.HashMap;

// https://leetcode.com/problems/two-sum/S

public class TwoSum {

    // Same as find_indexes but using HashMap

    public static int[] findTwoSum(int[] nums, int target) {
        // int[] ret = new int[2];

        // for (int i = 0; i < nums.length; i++) {
        // for (int j = i+1; j < nums.length; j++) {
        // if ((nums[i] + nums[j]) == target) {
        // ret[0] = i;
        // ret[1] = j;
        // }
        // }
        // }

        // return ret;

        HashMap<Integer, Integer> ret = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];

            if (ret.containsKey(nums[i])) {
                return new int[] { ret.get(num), i };
            }

            ret.put(target - num, i);
        }

        return new int[] {};

    }

    public static void printTwoSum(int[] arr) {
        System.out.print("[" + arr[0] + "," + arr[1] + "]\n");

    }

    public static void main(String[] args) {
        int[] arr1 = { 2, 7, 11, 15 };
        int[] arr1Ret = findTwoSum(arr1, 9);

        printTwoSum(arr1Ret);

        int[] arr2 = { 5, 7, 11, 15 };
        int[] arr2Ret = findTwoSum(arr2, 18);

        printTwoSum(arr2Ret);
    }
}
